sopht_mpi/numeric/eulerian_grid_ops/poisson_solver_3d/UnboundedPoissonSolverMPI3D.py

```diff
@@ -0,0 +1,378 @@
+            greens_function_field = (1 / even_reflected_distance_field) / (4 * np.pi)
+        # Regularization term (straight from PPM)
+        if np.all(global_start_idx == 0):
+            greens_function_field[0, 0, 0] = 1 / (4 * np.pi * self.dx)
+
+        # take forward transform of greens function field
+        self.rfft(
+            field=greens_function_field,
+            fourier_field=self.domain_doubled_fourier_buffer,
+        )
+
+    def gen_elementwise_operation_kernels(self):
+        """Compile funcs for elementwise ops on buffers."""
+        # this operate on domain doubled arrays
+        self.set_fixed_val_kernel_3d = gen_set_fixed_val_pyst_kernel_3d(
+            real_t=self.real_t
+        )
+        # this one operates on fourier buffer
+        self.elementwise_complex_product_kernel_3d = (
+            gen_elementwise_complex_product_pyst_kernel_3d(real_t=self.real_t)
+        )
+
+    def solve(self, solution_field, rhs_field):
+        """Unbounded Poisson solver method.
+
+        Solves Poisson equation in 3D: -del^2(solution_field) = rhs_field
+        for unbounded domain using Greens function convolution and
+        domain doubling trick (Hockney and Eastwood).
+        """
+
+        self.set_fixed_val_kernel_3d(field=self.domain_doubled_buffer, fixed_val=0)
+        self.mpi_domain_doubling_comm.copy_to_doubled_domain(
+            local_field=rhs_field,
+            local_doubled_field=self.domain_doubled_buffer,
+        )
+
+        self.rfft(
+            field=self.domain_doubled_buffer,
+            fourier_field=self.domain_doubled_fourier_buffer,
+        )
+
+        # Greens function convolution
+        self.elementwise_complex_product_kernel_3d(
+            product_field=self.convolution_buffer,
+            field_1=self.domain_doubled_fourier_buffer,
+            field_2=self.fourier_greens_function_times_dx_cubed,
+        )
+
+        self.irfft(
+            fourier_field=self.convolution_buffer,
+            inv_fourier_field=self.domain_doubled_buffer,
+        )
+
+        self.mpi_domain_doubling_comm.copy_from_doubled_domain(
+            local_doubled_field=self.domain_doubled_buffer,
+            local_field=solution_field,
+        )
+
+    def vector_field_solve(self, solution_vector_field, rhs_vector_field):
+        """Unbounded Poisson solver method (vector field solve).
+
+        Solves 3 Poisson equations in 3D for each component:
+        -del^2(solution_vector_field) = rhs_vector_field for unbounded domain using
+        Greens function convolution and domain doubling trick (Hockney and Eastwood).
+        """
+        self.solve(
+            solution_field=solution_vector_field[0], rhs_field=rhs_vector_field[0]
+        )
+        self.solve(
+            solution_field=solution_vector_field[1], rhs_field=rhs_vector_field[1]
+        )
+        self.solve(
+            solution_field=solution_vector_field[2], rhs_field=rhs_vector_field[2]
+        )
+
```

Comment:
vector indices

Reply:
Fixed in [c5a5c41](https://github.com/fankiat/sopht-mpi/pull/135/commits/c5a5c4170b4659e61254f22d4080f5b62c41172a)